kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/utils/TestUtils.scala

```diff
@@ -0,0 +1,244 @@
+    // TODO serialize scan state and scan rows
+
+    fileIter.forEach { fileColumnarBatch =>
+      // TODO deserialize scan state and scan rows
+      val dataBatches = Scan.readData(
+        tableClient,
+        scanState,
+        fileColumnarBatch.getRows(),
+        Optional.empty()
+      )
+
+      dataBatches.forEach { batch =>
+        val selectionVector = batch.getSelectionVector()
+        val data = batch.getData()
+
+        var i = 0
+        val rowIter = data.getRows()
+        try {
+          while (rowIter.hasNext) {
+            val row = rowIter.next()
+            if (!selectionVector.isPresent || selectionVector.get.getBoolean(i)) { // row is valid
+              result.append(row)
+            }
+            i += 1
+          }
+        } finally {
+          rowIter.close()
+        }
+      }
+    }
+    result
+  }
+
+  def withTimeZone(zoneId: String)(f: => Unit): Unit = {
+    val currentDefault = TimeZone.getDefault
+    try {
+      TimeZone.setDefault(TimeZone.getTimeZone(zoneId))
+      f
+    } finally {
+      TimeZone.setDefault(currentDefault)
+    }
+  }
+
+  def checkTable(
+    path: String,
+    expectedAnswer: Seq[TestRow],
+    readCols: Seq[String] = null,
+    tableClient: TableClient = defaultTableClient,
+    expectedSchema: StructType = null
+    // filter
+    // version
+  ): Unit = {
+
+    val snapshot = Table.forPath(path).getLatestSnapshot(tableClient)
+
+    val readSchema = if (readCols == null) {
+      null
+    } else {
+      val schema = snapshot.getSchema(tableClient)
+      new StructType(readCols.map(schema.get(_)).asJava)
+    }
+
+    if (expectedSchema != null) {
+      assert(
+        expectedSchema == snapshot.getSchema(tableClient),
+        s"""
+           |Expected schema does not match actual schema:
+           |Expected schema: $expectedSchema
+           |Actual schema: ${snapshot.getSchema(tableClient)}
+           |""".stripMargin
+      )
+    }
+
+    val result = readSnapshot(snapshot, readSchema = readSchema, tableClient = tableClient)
+    checkAnswer(result, expectedAnswer)
+  }
+
+  def checkAnswer(result: => Seq[Row], expectedAnswer: Seq[TestRow]): Unit = {
+    checkAnswer(result.map(TestRow(_)), expectedAnswer)
+  }
+
+  def checkAnswer(result: Seq[TestRow], expectedAnswer: Seq[TestRow]): Unit = {
+    if (!compare(prepareAnswer(result), prepareAnswer(expectedAnswer))) {
+      fail(genErrorMessage(expectedAnswer, result))
+    }
+  }
+
+  def prepareAnswer(answer: Seq[TestRow]): Seq[TestRow] = {
+    // Converts data to types that we can do equality comparison using Scala collections.
+    // For BigDecimal type, the Scala type has a better definition of equality test (similar to
+    // Java's java.math.BigDecimal.compareTo).
+    // For binary arrays, we convert it to Seq to avoid of calling java.util.Arrays.equals for
+    // equality test.
+    val converted = answer.map(prepareRow)
+    converted.sortBy(_.toString())
```

Comment:
Is a way to differentiate between the `long` and the `int` values which failed the comparison (expected a long value, but got int from kernel). In the printed error message it will look the same and can't tell what is wrong.

Reply:
Yeah that's what the TODO [here](https://github.com/delta-io/delta/pull/2034/files/7905c4d8e2d3c66d1a78e37c05cb6b02211cf9aa#diff-4b532e689ace6208898c69a5aa728bb7e76f97d08a39cc1632946c59f73f56a3R232) is for; was trying to time-box this as it got more complex on what the best way to do that was